src/main/java/io/aiven/kafka/connect/common/output/SinkSchemaBuilder.java

```diff
@@ -0,0 +1,183 @@
+    private final AvroData avroData;
+
+    private final boolean envelopeEnabled;
+
+    public SinkSchemaBuilder(final Collection<OutputField> fields,
+                      final AvroData avroData,
+                      final boolean envelopeEnabled) {
+        this.fields = fields;
+        this.avroData = avroData;
+        this.envelopeEnabled = envelopeEnabled;
+    }
+
+    public SinkSchemaBuilder(final Collection<OutputField> fields,
+                                final AvroData avroData) {
+        this.fields = fields;
+        this.avroData = avroData;
+        this.envelopeEnabled = true;
+    }
+
+    protected abstract String getNamespace();
+
+    public Schema buildSchema(final SinkRecord record) {
+        Objects.requireNonNull(record, "record");
+        if (Objects.isNull(record.keySchema())) {
+            throw new DataException("Record key without schema");
+        }
+        if (Objects.isNull(record.valueSchema())) {
+            throw new DataException("Record value without schema");
+        }
+        logger.debug("Create schema for record");
+        logger.debug("Record Key Schema {}", record.keySchema());
+        logger.debug("Record Value Schema {}", record.valueSchema());
+        return avroSchemaFor(record);
+    }
+
+    protected Schema avroSchemaFor(final SinkRecord record) {
+        if (envelopeEnabled) {
+            final SchemaBuilder.FieldAssembler<Schema> schemaFields =
+                SchemaBuilder
+                    .builder(getNamespace())
+                    .record("connector_records")
+                    .fields();
+            for (final OutputField f : fields) {
+                final Schema schema = outputFieldSchema(f, record);
+                schemaFields.name(f.getFieldType().name).type(schema).noDefault();
+            }
+            return schemaFields.endRecord();
+        } else {
+            return tryUnwrapEnvelope(record);
+        }
+    }
+
+    private Schema tryUnwrapEnvelope(final SinkRecord record) {
+        // envelope can be disabled only in case of single field
+        final OutputField field = getFields().iterator().next();
+
+        final Schema schema = outputFieldSchema(field, record);
+        if (schema.getType() == Schema.Type.MAP) {
+            @SuppressWarnings("unchecked") final Map<String, Object> value =
+                (Map<String, Object>) record.value();
+            final SchemaBuilder.FieldAssembler<Schema> schemaFields =
+                SchemaBuilder
+                    .builder(getNamespace())
+                    .record("connector_records")
+                    .fields();
+            for (final Map.Entry<String, Object> entry : value.entrySet()) {
+                schemaFields.name(entry.getKey()).type(schema.getValueType()).noDefault();
+            }
+            return schemaFields.endRecord();
+        } else if (schema.getType() == Schema.Type.RECORD) {
+            return getAvroData().fromConnectSchema(record.valueSchema());
```

Comment:
I see we are assuming the single field is value (e.g. `record.value()` `record.valueSchema()`). Should we do these after checking that the field is actually `VALUE`?
Then, we can remove the `else` and create the schema builder in any other case.